File: apps/tone-embedding/src/tone_embedding/taxonomy.py
from __future__ import annotations

import json
from functools import lru_cache
from pathlib import Path
from typing import Any


TONE_TAXONOMY_VERSION = "tone-taxonomy/v1"
TAXONOMY_DIR = Path(__file__).resolve().parent / "taxonomies"


@lru_cache(maxsize=1)
def load_tone_taxonomy() -> dict[str, Any]:
    path = TAXONOMY_DIR / "tone-taxonomy.v1.json"
    with path.open("r", encoding="utf-8") as file:
        taxonomy = json.load(file)
    validate_tone_taxonomy(taxonomy)
    return taxonomy


def validate_tone_taxonomy(taxonomy: dict[str, Any]) -> None:
    if taxonomy.get("schemaVersion") != TONE_TAXONOMY_VERSION:
        raise ValueError(f"unsupported tone taxonomy version: {taxonomy.get('schemaVersion')}")
    if not isinstance(taxonomy.get("dimensions"), list) or not taxonomy["dimensions"]:
        raise ValueError("tone taxonomy dimensions must be a non-empty array")
    if not isinstance(taxonomy.get("toneDimensions"), list) or not taxonomy["toneDimensions"]:
        raise ValueError("tone taxonomy toneDimensions must be a non-empty array")
    if not isinstance(taxonomy.get("descriptors"), list) or not taxonomy["descriptors"]:
        raise ValueError("tone taxonomy descriptors must be a non-empty array")
    if not isinstance(taxonomy.get("strengthScale"), dict):
        raise ValueError("tone taxonomy strengthScale must be an object")
    if not isinstance(taxonomy.get("avoidRules"), dict):
        raise ValueError("tone taxonomy avoidRules must be an object")
    dimension_ids = {str(dimension["id"]) for dimension in taxonomy["dimensions"]}
    vector_dimensions = {str(dimension) for dimension in taxonomy["toneDimensions"]}
    if dimension_ids != vector_dimensions:
        raise ValueError("tone taxonomy dimensions and toneDimensions must contain the same ids")


def taxonomy_dimensions(taxonomy: dict[str, Any] | None = None) -> tuple[str, ...]:
    taxonomy = taxonomy or load_tone_taxonomy()
    return tuple(str(dimension) for dimension in taxonomy["toneDimensions"])


def descriptor_rules(taxonomy: dict[str, Any] | None = None) -> tuple[tuple[str, float, str, str], ...]:
    taxonomy = taxonomy or load_tone_taxonomy()
    return tuple(
        (
            str(dimension["id"]),
            float(dimension["descriptorThreshold"]),
            str(dimension["positiveDescriptor"]),
            str(dimension["negativeDescriptor"]),
        )
        for dimension in taxonomy["dimensions"]
    )


def strength_scores(taxonomy: dict[str, Any] | None = None) -> dict[str, float]:
    taxonomy = taxonomy or load_tone_taxonomy()
    return {str(name): float(value) for name, value in taxonomy["strengthScale"].items()}


def avoid_rules(taxonomy: dict[str, Any] | None = None) -> dict[str, str]:
    taxonomy = taxonomy or load_tone_taxonomy()
    return {str(name): str(value) for name, value in taxonomy["avoidRules"].items()}


def descriptor_to_score(taxonomy: dict[str, Any] | None = None) -> dict[str, tuple[str, float]]:
    return {
        positive_word: (dimension, 1.0)
        for dimension, _, positive_word, _ in descriptor_rules(taxonomy)
    } | {
        negative_word: (dimension, -1.0)
        for dimension, _, _, negative_word in descriptor_rules(taxonomy)
    }


def upgrade_tone_taxonomy(payload: dict[str, Any], target_version: str) -> dict[str, Any]:
    raise NotImplementedError(f"tone taxonomy upgrade to {target_version} is not implemented")


def downgrade_tone_taxonomy(payload: dict[str, Any], target_version: str) -> dict[str, Any]:
    raise NotImplementedError(f"tone taxonomy downgrade to {target_version} is not implemented")

File: packages/tone-core/src/taxonomy.ts
import taxonomy from "./taxonomies/tone-taxonomy.v2.json" with { type: "json" };
import {
  TONE_TAXONOMY_VERSION,
  ToneDimensionSchema,
  type StructuredToneDescriptor,
  type ToneDimension,
} from "./schemas.js";

type RawTaxonomyDimension = {
  id: string;
  descriptorThreshold: number;
  positiveDescriptor: string;
  negativeDescriptor: string;
};

export type DescriptorMapping = {
  dimension: ToneDimension;
  weight: number;
  confidence?: number;
  basis?: string;
};

type RawTaxonomyDescriptor = {
  slug: string;
  status: string;
  mappings?: Array<{
    dimension: string;
    weight: number;
    confidence?: number;
    basis?: string;
  }>;
};

type RawToneTaxonomy = {
  schemaVersion: string;
  toneDimensions: string[];
  dimensions: RawTaxonomyDimension[];
  descriptors: RawTaxonomyDescriptor[];
  strengthScale: Record<string, number>;
  avoidRules: Record<string, string>;
};

const toneTaxonomy = taxonomy as RawToneTaxonomy;

export function loadToneTaxonomy(): RawToneTaxonomy {
  if (toneTaxonomy.schemaVersion !== TONE_TAXONOMY_VERSION) {
    throw new Error(`Unsupported tone taxonomy version: ${toneTaxonomy.schemaVersion}`);
  }
  return toneTaxonomy;
}

export function toneDimensions(): ToneDimension[] {
  return loadToneTaxonomy().toneDimensions.map((dimension) => ToneDimensionSchema.parse(dimension));
}

export function descriptorRules(): Array<{
  dimension: ToneDimension;
  threshold: number;
  positiveDescriptor: string;
  negativeDescriptor: string;
}> {
  return loadToneTaxonomy().dimensions.map((dimension) => ({
    dimension: ToneDimensionSchema.parse(dimension.id),
    threshold: dimension.descriptorThreshold,
    positiveDescriptor: dimension.positiveDescriptor,
    negativeDescriptor: dimension.negativeDescriptor,
  }));
}

export function strengthScores(): Record<string, number> {
  return loadToneTaxonomy().strengthScale;
}

export function avoidRules(): Record<string, string> {
  return loadToneTaxonomy().avoidRules;
}

export function descriptorMappings(): Record<string, DescriptorMapping[]> {
  const output: Record<string, DescriptorMapping[]> = {};
  for (const descriptor of loadToneTaxonomy().descriptors) {
    if (descriptor.status !== "active" || !descriptor.mappings || descriptor.mappings.length === 0) {
      continue;
    }
    output[descriptor.slug] = descriptor.mappings.map((mapping) => ({
      dimension: ToneDimensionSchema.parse(mapping.dimension),
      weight: mapping.weight,
      confidence: mapping.confidence,
      basis: mapping.basis,
    }));
  }
  return output;
}

export function supportedDescriptors(): string[] {
  return Object.keys(descriptorMappings());
}

export function normalizeDescriptor(
  descriptor: StructuredToneDescriptor
): StructuredToneDescriptor {
  const mappings = descriptorMappings()[descriptor.descriptor];
  if (!mappings) {
    throw new Error(`Unsupported tone descriptor: ${descriptor.descriptor}`);
  }
  if (!descriptor.dimension) {
    return descriptor;
  }
  return {
    ...descriptor,
    mappedDimension: mappings[0]?.dimension,
    modelDimension: descriptor.dimension,
  };
}
